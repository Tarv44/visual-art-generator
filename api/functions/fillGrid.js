const generateEmptyGrid = require('./generateEmptyGrid');
const { randomFromRange, randomFromArray } = require('./helperFunctions');
const generateParams =  require('./generate-params');

function selectRandomCell(totalColumns, totalRows) {
    const column = randomFromRange(totalColumns);
    const row = randomFromRange(totalRows);
    return {column, row}
}

//Generates random rgb value based on color range constraints
function rangedRandomRGB(gridConst) {
    const colorRanges = gridConst.skew
    const colorObject = {}
    colorObject.red = (randomFromRange(colorRanges.redUpper - colorRanges.redLow + 1) + colorRanges.redLow)
    colorObject.green = (randomFromRange(colorRanges.greenUpper - colorRanges.greenLow + 1) + colorRanges.greenLow)
    colorObject.blue = (randomFromRange(colorRanges.blueUpper - colorRanges.blueLow + 1) + colorRanges.blueLow)
    const rgb = `rgba(${colorObject.red}, ${colorObject.green}, ${colorObject.blue}, 1)`
    colorObject.rgb = rgb
    return colorObject
}

//Given a single cell's coordinates, 
//retrieves coordinates of all surrounding cells within the grid
function allSurrCoors(gridCols, cellCoor) {
    const columns = [cellCoor.column-1, cellCoor.column, cellCoor.column+1]
        .filter(num => num > -1 && num < gridCols.length)
    const rows = [cellCoor.row-1, cellCoor.row, cellCoor.row+1]
        .filter(num => num > -1 && num < gridCols[0].length)
    const coors = [];
    for (let c = 0; c < columns.length; c++) {
        for (let r = 0; r < rows.length; r++) {
            const id = `c${columns[c]}r${rows[r]}`
            const coor = {column: columns[c], row: rows[r], id}
            if (!(cellCoor.column === coor.column && cellCoor.row === coor.row)) {
                coors.push(coor)
            }

        }
    }
    return coors
}

//Filters all surrounding cells coordinates so that only cells
//with no color are returned.
function getSurrEmpties(gridCols, cellCoor) {
    const surrCoors = allSurrCoors(gridCols, cellCoor)
    const surrEmpties = surrCoors.filter(coor => 
        gridCols[coor.column][coor.row].color === null)
    return surrEmpties
}

//Filters all surrounding cells coordinates so that only cells
//with colors are returned.
function getSurrColors(gridCols, cellCoor) {
    const coors = allSurrCoors(gridCols, cellCoor);
    const surrColors = coors.filter(coor =>
        gridCols[coor.column][coor.row].color !== null)
    return surrColors
}

//Updates fillable cells array with new fillable cells and removes
//cell that is currently being filled
function updateFillableCells(gridCols, fillableCells, currentCoor) {
    const newList = fillableCells.filter(cell => cell.id !== currentCoor.id)
    const allSurrEmpties = getSurrEmpties(gridCols, currentCoor)
    const newFillables = []
    for(let i = 0; i < allSurrEmpties.length; i++) {
        if(!fillableCells.some(cell => cell.id === allSurrEmpties[i].id)) {
            newFillables.push(allSurrEmpties[i])
        }
    }
    return newList.concat(newFillables)
}

//Given values for r, g, and b, select new random values that
//are within the given skew range.
//I.e. red = 110, skew = +- 10, new red returned = 117. 
function skewColor(color, currentGridConstraints) {
    const range = currentGridConstraints.skew.changeRange
    let red 
    let green
    let blue
    const redChange = (randomFromRange(range) + 1)
    const greenChange = (randomFromRange(range) + 1)
    const blueChange = (randomFromRange(range) + 1)

    if (Math.random() < .5) {
        red = color.red + redChange
        if (red > currentGridConstraints.skew.redUpper) {
            red = currentGridConstraints.skew.redUpper
        }
    } else {
        red = color.red - redChange
        if (red < currentGridConstraints.skew.redLow) {
            red = currentGridConstraints.skew.redLow
        }
    }

    if (Math.random() < .5) {
        green = color.green + greenChange
        if (green > currentGridConstraints.skew.greenUpper) {
            green = currentGridConstraints.skew.greenUpper
        }
    } else {
        green = color.green - greenChange
        if (green < currentGridConstraints.skew.greenLow) {
            green = currentGridConstraints.skew.greenLow
        }
    }

    if (Math.random() < .5) {
        blue = color.blue + blueChange
        if (blue > currentGridConstraints.skew.blueUpper) {
            blue = currentGridConstraints.skew.blueUpper
        }
    } else {
        blue = color.blue - blueChange
        if (blue < currentGridConstraints.skew.blueLow) {
            blue = currentGridConstraints.skew.blueLow
        }
    }

    const rgb = `rgba(${red}, ${green}, ${blue}, 1)`

    const newColor = {red, green, blue, rgb}
    return newColor
}

//Randomly select (given the predetermined probabilities) 
//whether the cell currently being filled will receive:
//A. The same exact color as a surrounding cell
//B. A skewed color based off the color of a surrounding cell
//C. A completely random color
function selectNewColor(surrColor, currentGridConstraints) {
    const probability = (Math.random()*100)
    const colorChances = currentGridConstraints.colorChances
    let newColor
    if (probability <= parseFloat(colorChances.same)) {
        newColor = surrColor
    } else if (probability <= parseFloat(colorChances.same) + parseFloat(colorChances.skew)) {
        newColor = skewColor(surrColor, currentGridConstraints)
    } else {
        newColor = rangedRandomRGB(currentGridConstraints)
    }
    return newColor
}


//Fill initial starting nodes of grid, given the predetermined
//amount of starting nodes.
function fillInitCells(currentGrid) {
    const totalStartNodes = currentGrid.params.node.totalStart
    for (let i = 1; i <= totalStartNodes; i++) {
        const cellCoor = selectRandomCell(currentGrid.totalColumns, currentGrid.totalRows);
        currentGrid.grid.columns[cellCoor.column][cellCoor.row].color = rangedRandomRGB(currentGrid.params);
        currentGrid.grid.columns[cellCoor.column][cellCoor.row].opacity = 1;
        currentGrid.totalCellsFilled += 1
        currentGrid.fillableCells = currentGrid.fillableCells.concat(getSurrEmpties(currentGrid.grid.columns, cellCoor))
    }
    return currentGrid
}

//If grid is generated or is completely full, generate new grid 
//and fill initial cells. Otherwise, simply hide form for restart.
function fillStart(totalColumns, totalRows, cellSize) {
    const gridData = {
        totalColumns,
        totalRows,
        totalCells: totalRows * totalColumns,
        cellSize,
        totalCellsFilled: 0,
        fillableCells: [],
    }

    gridData.params = generateParams()
    gridData.params.node = {totalStart: 1}
    gridData.grid = generateEmptyGrid(totalColumns, totalRows)
    gridStart = fillInitCells(gridData)

    return gridStart
}

//Go through process of selecting a random fillable cell
//and selecting a color for it.
function fillColor(currentGrid) {
    const gridCols = currentGrid.grid.columns
    const fillableCoor = randomFromArray(currentGrid.fillableCells)
    const fillableCells = updateFillableCells(gridCols, currentGrid.fillableCells, fillableCoor)
    const surrColorsCoors = getSurrColors(gridCols, fillableCoor)
    const baseColorCoor = randomFromArray(surrColorsCoors)
    const baseColor = currentGrid.grid.columns[baseColorCoor.column][baseColorCoor.row].color
    const newColor = selectNewColor(baseColor, currentGrid.params)
    currentGrid.grid.columns[fillableCoor.column][fillableCoor.row].color = newColor;
    currentGrid.grid.columns[fillableCoor.column][fillableCoor.row].opacity = 1;
    currentGrid.totalCellsFilled += 1;
    currentGrid.fillableCells = fillableCells

    return currentGrid
}

//Fill grid until the designated portion of grid to be filled
//is reached, or until grid is full, whichever comes first.
function fillCellGroup(currentGrid) {
    let newGrid = currentGrid
    
    while (newGrid.totalCells !== newGrid.totalCellsFilled) {
        newGrid = fillColor(newGrid)
    }
    
    return newGrid
}

module.exports = {
    fillStart,
    fillCellGroup
}