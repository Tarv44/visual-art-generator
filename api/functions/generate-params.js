const {randomFromRange} = require('./helperFunctions');

const generateColorChances = () => {
    const colorChances = {}
    
    //Chance of random color
    const randomChance = randomFromRange(3)
    if (randomChance === 0) {
        colorChances.random = 0
    }

    if (randomChance === 1) {
        colorChances.random = (randomFromRange(100) + 1)/100
    }

    if (randomChance === 2) {
        colorChances.random = (randomFromRange(600) + 1)/100
    }

    //Chance of Skew Color
    const skewChance = randomFromRange(6)
    if (skewChance === 0) {
        colorChances.skew = 0
    }

    if (skewChance >= 1 && skewChance <= 3) {
        colorChances.skew = (Math.random() * (100 - colorChances.random)) + 1
    }

    if (skewChance >= 4) {
        colorChances.skew = 100 - colorChances.random
    }

    //Chance of Same Color
    colorChances.same = 100 - colorChances.random - colorChances.skew

    return colorChances
}

const generateSkewParams = () => {
    const skewParams = {}
    
    //Skew Range
    const skewRangeChance = randomFromRange(4)
    if (skewRangeChance === 0) {
        skewParams.changeRange = randomFromRange(3) + 1
    }

    if (skewRangeChance === 1) {
        skewParams.changeRange = randomFromRange(4) + 4
    }

    if (skewRangeChance === 2) {
        skewParams.changeRange = randomFromRange(8) + 9
    }

    if (skewRangeChance === 3) {
        skewParams.changeRange = randomFromRange(22) + 18
    }

    //Color Range
    const colorRangeChance = randomFromRange(100)

    if (colorRangeChance === 0) {
        skewParams.redLow = 0
        skewParams.redUpper = 255
        skewParams.blueLow = 0
        skewParams.blueUpper = 255
        skewParams.greenLow = 0
        skewParams.greenUpper = 255
    }

    if (colorRangeChance > 0) {
        //Red
        if (randomFromRange(100) < 5) {
            skewParams.redLow = 0
            skewParams.redUpper = 255
        } else {
            const num1 = randomFromRange(256)
            const num2 = randomFromRange(256)
            if (num1 >= num2) {
                skewParams.redUpper = num1
                skewParams.redLow = num2
            } else {
                skewParams.redUpper = num2
                skewParams.redLow = num1
            }
        }

        //Blue
        if (randomFromRange(100) < 5) {
            skewParams.blueLow = 0
            skewParams.blueUpper = 255
        } else {
            const num1 = randomFromRange(256)
            const num2 = randomFromRange(256)
            if (num1 >= num2) {
                skewParams.blueUpper = num1
                skewParams.blueLow = num2
            } else {
                skewParams.blueUpper = num2
                skewParams.blueLow = num1
            }
        }

        //Green
        if (randomFromRange(100) < 5) {
            skewParams.greenLow = 0
            skewParams.greenUpper = 255
        } else {
            const num1 = randomFromRange(256)
            const num2 = randomFromRange(256)
            if (num1 >= num2) {
                skewParams.greenUpper = num1
                skewParams.greenLow = num2
            } else {
                skewParams.greenUpper = num2
                skewParams.greenLow = num1
            }
        }
    }

    return skewParams
}

// const generateNodeParams = () => {
//     const nodeParams = {}

//     //Starting nodes
//     const startNodeChance = randomFromRange(3)

//     if (startNodeChance === 0) {
//         nodeParams.totalStart = 1
//     }

//     if (startNodeChance === 1) {
//         nodeParams.totalStart = randomFromRange(10) + 1
//     }

//     if (startNodeChance === 2) {
//         nodeParams.totalStart = randomFromRange(100) + 1
//     }

//     // if (startNodeChance === 3) {
//     //     nodeParams.totalStart = randomFromRange(1000) + 1
//     // }

//     nodeParams.chanceNew = 0
//     nodeParams.chanceNewSame = 0
//     nodeParams.chanceNewDiff = 0

//     return nodeParams
// }

function generateParams() {
    const params = {}
    params.colorChances = generateColorChances()
    params.skew = generateSkewParams()
    // newGrid.params.node = generateNodeParams()
    return params
} 

module.exports = generateParams
