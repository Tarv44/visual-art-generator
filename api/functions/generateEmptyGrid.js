function generateEmptyCell(col, r) {
    const cellId = `c${col}r${r}`
    const emptyCell = {
        id: cellId,
        column: col,
        row: r,
        opacity: 0,
        color: null,
    }
    return emptyCell
}

function generateEmptyGrid(totalColumns, totalRows) {
    const columns = []
    for (let c = 0; c < totalColumns; c++) {
        const columnArray = []
        for (let r = 0; r < totalRows; r++) {
            columnArray.push(generateEmptyCell(c, r))
        }
        columns.push(columnArray)
    }
    return {columns}
}

function generateDummyGrid() {
    const gridState = {}
    const gridArray = []
    for (let c = 0; c < 10; c++) {
        const columnArray = []
        for (let r = 0; r < 10; r++) {
            columnArray.push(generateEmptyCell(c, r))
        }
        gridArray.push(columnArray)
    }
    gridState.columns = gridArray
    return gridState
}

module.exports = generateEmptyGrid