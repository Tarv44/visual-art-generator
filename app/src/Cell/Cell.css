.cell {
    height: 15px;
    width: 15px;
    /* transition: 1s ease-in opacity; */
}

@keyframes fadeIn {
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
}