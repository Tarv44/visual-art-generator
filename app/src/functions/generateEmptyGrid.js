function generateEmptyCell(col, r, size) {
    const cellId = `c${col}r${r}`
    const emptyCell = {
        id: cellId,
        column: col,
        row: r,
        opacity: 0,
        color: null,
        size
    }
    return emptyCell
}

export default function generateEmptyGrid(state) {
    const gridState = state
    const columns = []
    for (let c = 0; c < state.totalColumns; c++) {
        const columnArray = []
        for (let r = 0; r < state.totalRows; r++) {
            columnArray.push(generateEmptyCell(c, r, state.cellSize))
        }
        columns.push(columnArray)
    }
    gridState.grid = {columns}
    gridState.totalCellsFilled = 0
    return gridState
}

export function generateDummyGrid() {
    const gridState = {}
    const gridArray = []
    for (let c = 0; c < 10; c++) {
        const columnArray = []
        for (let r = 0; r < 10; r++) {
            columnArray.push(generateEmptyCell(c, r))
        }
        gridArray.push(columnArray)
    }
    gridState.columns = gridArray
    return gridState
}