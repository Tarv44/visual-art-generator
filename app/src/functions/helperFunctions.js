export function randomFromRange(range) {
    return Math.floor(Math.random() * range)
}

export function randomFromArray(array) {
    const i = randomFromRange(array.length)
    return array[i]
}

export function randomRGB() {
    const colorObject = {}
    colorObject.red = randomFromRange(256)
    colorObject.green = randomFromRange(256)
    colorObject.blue = randomFromRange(256)
    const rgb = `rgba(${colorObject.red}, ${colorObject.green}, ${colorObject.blue}, 1)`
    colorObject.rgb = rgb
    return colorObject
}