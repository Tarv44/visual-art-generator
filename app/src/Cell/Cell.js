import React, {Component} from 'react';
import './Cell.css'

export default class Cell extends Component {
    static defaultProps = {
        cellData: {
            id: '',
            column: 0,
            row: 0,
            opacity: 0,
            color: {
                red: 0,
                green: 0,
                blue: 0,
                rgb: ''
            }
        }
    }

    render() {
        // const cellColor = this.props.cellData.color ? this.props.cellData.color.rgb : '';
        // const animationDelay = this.props.cellData.animationDelay + 's'
        const styles = {
            backgroundColor: this.props.cellData.color.rgb,
            width: this.props.cellData.size,
            height: this.props.cellData.size
        };
        return (
            <div style={styles} className={'cell'}/>
        )
    }
}