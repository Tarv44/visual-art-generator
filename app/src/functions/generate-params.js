import {randomFromRange as random} from './helperFunctions';

const generateColorChances = () => {
    const colorChances = {}
    
    //Chance of random color
    const randomChance = random(3)
    if (randomChance === 0) {
        colorChances.random = 0
    }

    if (randomChance === 1) {
        colorChances.random = (random(100) + 1)/100
    }

    if (randomChance === 2) {
        colorChances.random = (random(600) + 1)/100
    }

    //Chance of Skew Color
    const skewChance = random(6)
    if (skewChance === 0) {
        colorChances.skew = 0
    }

    if (skewChance >= 1 && skewChance <= 3) {
        colorChances.skew = (Math.random() * (100 - colorChances.random)) + 1
    }

    if (skewChance >= 4) {
        colorChances.skew = 100 - colorChances.random
    }

    //Chance of Same Color
    colorChances.same = 100 - colorChances.random - colorChances.skew

    return colorChances
}

const generateSkewParams = () => {
    const skewParams = {}
    
    //Skew Range
    const skewRangeChance = random(4)
    if (skewRangeChance === 0) {
        skewParams.changeRange = random(3) + 1
    }

    if (skewRangeChance === 1) {
        skewParams.changeRange = random(4) + 4
    }

    if (skewRangeChance === 2) {
        skewParams.changeRange = random(8) + 9
    }

    if (skewRangeChance === 3) {
        skewParams.changeRange = random(22) + 18
    }

    //Color Range
    const colorRangeChance = random(100)

    if (colorRangeChance === 0) {
        skewParams.redLow = 0
        skewParams.redUpper = 255
        skewParams.blueLow = 0
        skewParams.blueUpper = 255
        skewParams.greenLow = 0
        skewParams.greenUpper = 255
    }

    if (colorRangeChance > 0) {
        //Red
        if (random(100) < 5) {
            skewParams.redLow = 0
            skewParams.redUpper = 255
        } else {
            const num1 = random(256)
            const num2 = random(256)
            if (num1 >= num2) {
                skewParams.redUpper = num1
                skewParams.redLow = num2
            } else {
                skewParams.redUpper = num2
                skewParams.redLow = num1
            }
        }

        //Blue
        if (random(100) < 5) {
            skewParams.blueLow = 0
            skewParams.blueUpper = 255
        } else {
            const num1 = random(256)
            const num2 = random(256)
            if (num1 >= num2) {
                skewParams.blueUpper = num1
                skewParams.blueLow = num2
            } else {
                skewParams.blueUpper = num2
                skewParams.blueLow = num1
            }
        }

        //Green
        if (random(100) < 5) {
            skewParams.greenLow = 0
            skewParams.greenUpper = 255
        } else {
            const num1 = random(256)
            const num2 = random(256)
            if (num1 >= num2) {
                skewParams.greenUpper = num1
                skewParams.greenLow = num2
            } else {
                skewParams.greenUpper = num2
                skewParams.greenLow = num1
            }
        }
    }

    return skewParams
}

// const generateNodeParams = () => {
//     const nodeParams = {}

//     //Starting nodes
//     const startNodeChance = random(3)

//     if (startNodeChance === 0) {
//         nodeParams.totalStart = 1
//     }

//     if (startNodeChance === 1) {
//         nodeParams.totalStart = random(10) + 1
//     }

//     if (startNodeChance === 2) {
//         nodeParams.totalStart = random(100) + 1
//     }

//     // if (startNodeChance === 3) {
//     //     nodeParams.totalStart = random(1000) + 1
//     // }

//     nodeParams.chanceNew = 0
//     nodeParams.chanceNewSame = 0
//     nodeParams.chanceNewDiff = 0

//     return nodeParams
// }

export default function generateParams(currentGrid) {
    const newGrid = currentGrid
    newGrid.params.colorChances = generateColorChances()
    newGrid.params.skew = generateSkewParams()
    // newGrid.params.node = generateNodeParams()
    return newGrid
} 
