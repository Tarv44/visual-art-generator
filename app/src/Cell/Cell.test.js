import React, { createElement } from 'react';
import ReactDom from 'react-dom';
import Cell from './Cell';
import {randomRGB} from '../functions/colorFunctions';
import renderer from 'react-test-renderer';

const dummyCell = {
    id: 'c1r1',
    column: 1,
    row: 1,
    gridSize: 'small',
    animationDelay: 1,
    color: {
        red: 0,
        green: 0,
        blue: 0,
        rgb: 'rgba(0, 0, 0, 1)'
    }
}

it('renders without crashing',() => {
    const div = document.createElement('div')
    ReactDom.render(<Cell />, div)
    ReactDom.unmountComponentAtNode(div)
})

it('renders UI as expected', () => {
    const cell = renderer
        .create(<Cell cellData={dummyCell}/>)
        .toJSON()
    expect(cell).toMatchSnapshot();
})