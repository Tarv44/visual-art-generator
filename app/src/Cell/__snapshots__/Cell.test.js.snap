// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders UI as expected 1`] = `
<div
  className="cell small"
  style={
    Object {
      "animationDelay": "1s",
      "backgroundColor": "rgba(0, 0, 0, 1)",
    }
  }
/>
`;
